core/src/test/java/org/elasticsearch/bootstrap/MockPluginPolicy.java

```diff
@@ -0,0 +1,123 @@
+import java.security.ProtectionDomain;
+import java.security.cert.Certificate;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Simulates in unit tests per-plugin permissions.
+ * Unit tests for plugins do not have a proper plugin structure,
+ * so we don't know which codebases to apply the permission to.
+ * <p>
+ * As an approximation, we just exclude es/test/framework classes,
+ * because they will be present in stacks and fail tests for the 
+ * simple case where an AccessController block is missing.
+ */
+final class MockPluginPolicy extends Policy {
+    final ESPolicy standardPolicy;
+    final PermissionCollection extraPermissions;
+    final Set<CodeSource> extraSources;
+
+    /**
+     * Create a new MockPluginPolicy with dynamic {@code permissions} and
+     * adding the extra plugin permissions from {@code insecurePluginProp} to
+     * all code except test classes.
+     */
+    MockPluginPolicy(Permissions permissions, String insecurePluginProp) throws Exception {
+        // the hack begins!
+
+        // parse whole policy file, with and without the substitution, compute the delta
+        standardPolicy = new ESPolicy(permissions);
+
+        URL bogus = new URL("file:/bogus"); // its "any old codebase" this time: generic permissions
+        PermissionCollection smallPermissions = standardPolicy.template.getPermissions(new CodeSource(bogus, (Certificate[])null)); 
+        Set<Permission> small = new HashSet<>(Collections.list(smallPermissions.elements()));
+
+        // set the URL for the property substitution, this time it will also have special permissions
+        System.setProperty(insecurePluginProp, bogus.toString());
+        ESPolicy biggerPolicy = new ESPolicy(permissions);
+        System.clearProperty(insecurePluginProp);
+        PermissionCollection bigPermissions = biggerPolicy.template.getPermissions(new CodeSource(bogus, (Certificate[])null));
+        Set<Permission> big = new HashSet<>(Collections.list(bigPermissions.elements()));
+
+        // compute delta to remove all the generic permissions
+        // we want equals() vs implies() for this check, in case we need 
+        // to pass along any UnresolvedPermission to the plugin
+        big.removeAll(small);
+
+        // build collection of the special permissions for easy checking
+        extraPermissions = new Permissions();
+        for (Permission p : big) {
+            extraPermissions.add(p);
+        }
+
+        // every element in classpath except test-classes/
+        extraSources = new HashSet<CodeSource>();
+        for (URL location : JarHell.parseClassPath()) {
+            Path path = PathUtils.get(location.toURI());
+            String baseName = path.getFileName().toString();
+            if (baseName.contains("test-classes") == false) {
+              extraSources.add(new CodeSource(location, (Certificate[])null));
+            }
+        }
+        // exclude some obvious places
```

Comment:
Does this help? https://github.com/rmuir/elasticsearch/commit/0fdc16927efa89a98285fd3b9ba51d4f1a7cb1f2